src/legacy/core_plugins/timelion/public/components/timelion_expression_input.tsx

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { useEffect, useCallback, useRef } from 'react';
+import { EuiFormRow } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import * as monacoEditor from 'monaco-editor/esm/vs/editor/editor.api';
+
+import { IUiSettingsClient, HttpSetup } from 'kibana/public';
+import { CodeEditor, KibanaContextProvider } from '../../../../../plugins/kibana_react/public';
+import { suggest, getSuggestion } from './timelion_expression_input_helpers';
+import { ArgValueSuggestions } from '../services/arg_value_suggestions';
+import { ITimelionFunction, TimelionFunctionArgs } from '../../common/types';
+
+const LANGUAGE_ID = 'timelion_expression';
+monacoEditor.languages.register({ id: LANGUAGE_ID });
+
+export interface TimelionExpressionInputDependencies {
+  argValueSuggestions: ArgValueSuggestions;
+  http: HttpSetup;
```

Comment:
Why not use the services pattern used recently in many places?
https://github.com/elastic/kibana/blob/master/src/plugins/data/public/services.ts

It will simplify the code greatly.

Reply:
I do not think it is a good idea to use such a pattern everywhere..
But I've used the context pattern instead -> https://github.com/elastic/kibana/pull/52990#discussion_r359736320